### DISCUSSION TITLE
Get ReadStream from uploaded file

### DISCUSSION THREAD

**Discussion by keesvanbemmel** [directus/directus#17149] [Archive]
Hi all,

I'm trying to get a readstream from an uploaded file. Before the [Storage refactor](https://github.com/directus/directus/pull/16825) we achieved this like this:
```
const storage = require('directus/storage');
const disk = storage.default.disk(input.payload.storage);
const readStream = disk.getStream(input.payload.filename_disk);
```

where input is the input in a files.upload hook.

Now, obviously, the module 'directus/storage' is gone. But I'm having a hard time to replace/refactor it on our side.

Any thoughts?

Cheers!

**Comment by paescuj** [comment]
This is what I came up with on the rapid:
```js
	action('files.upload', async ({ payload }) => {
		const { getStorage } = require('directus/storage/index');
		const storage = await getStorage();
		const file = await storage.location(payload.storage).read(payload.filename_disk);
		// Do things with 'file'...
	});
```

**Note:** As you're probably already aware of, the use of exports from `directus` is not something which is officially supported and those may change at any time.

**Reply by keesvanbemmel** [reply]
Works like a charm! Thanks a bunch @paescuj !

**Reply by paescuj** [reply]
Glad it's working for you :-) You're welcome!

**Comment by beniw** [comment]
Hello everyone,

Sorry to bring this up again.

We used something similar in an earlier version of directus:

`const { default: storage } = require('/directus/node_modules/directus/dist/storage');`

This no longer seems to work after the v10.4.3 update.

How would you do this nowadays with directus v10.4.3?

Thank you very much in advance :)

**Reply by rijkvanzanten** [reply]
You might be able to pull it from `@directus/api` instead, but this hasn't been tested nor is officially supported, so your mileage may vary!

**Reply by beniw** [reply]
Thank you for your reply!

I tried to pull it from @directus/api but I ran into the following issue with require:

`const { default: getStorage } = require('/directus/node_modules/@directus/api')`

`This file is being treated as an ES module because it has a '.js' file extension and '/directus/extensions/hooks/test/package.json' contains "type": "module". To treat it as a CommonJS script, rename it to use the '.cjs' file extension.`


When I change it to an import I get this:

`import { getStorage } from '/directus/node_modules/@directus/api'`

`⠦ Building Directus extension...The 'this' keyword is equivalent to 'undefined' at the top level of an ES module, and has been rewritten
⠇ Building Directus extension..."fsevents" is imported by "../../../node_modules/.pnpm/rollup@3.22.0/node_modules/rollup/dist/es/shared/node-entry.js", but could not be resolved – treating it as an external dependency.
⠇ Building Directus extension...The 'this' keyword is equivalent to 'undefined' at the top level of an ES module, and has been rewritten
The 'this' keyword is equivalent to 'undefined' at the top level of an ES module, and has been rewritten
✖ Failed
[RollupError] Unexpected character '' (Note that you need plugins to import files that are not JavaScript) (plugin commonjs--resolver)`

Do you have any idea what I'm doing wrong? 

Thank you very much in advance :)

**Reply by paescuj** [reply]
There you go:
```js
	action('files.upload', async ({ payload }) => {
		const { getStorage } = await import('@directus/api/storage/index');
		const storage = await getStorage();
		const file = await storage.location(payload.storage).read(payload.filename_disk);
		// Do things with 'file'...
	});
```

**Reply by beniw** [reply]
Thanks  for the fast response @paescuj  🙌 

But now I'm getting this during the build:

`[SyntaxError] sleep is redeclared (plugin terser)`

**Reply by paescuj** [reply]
You're welcome!
Hmm, this seems to be unrelated. Would you mind posting the full code?

**Reply by beniw** [reply]
Yea I thought the same. 

I'll try again tomorrow with a fresh brain and post the full code if I can not get it to work.

Thanks and have a good one :)

**Reply by paescuj** [reply]
Okay 👍 Good luck & enjoy the weekend 🌞

**Reply by beniw** [reply]
Hi again,

I hope you had a nice weekend.

I tried again and got a different result. Currently I'm getting the following error message:

```
docker-cms-1     | node:internal/event_target:1012
docker-cms-1     |   process.nextTick(() => { throw err; });
docker-cms-1     |                            ^
docker-cms-1     | Error: Cannot find module '/directus/extensions/hooks/test-file-upload/dist/lib/worker.js'
docker-cms-1     |     at Module._resolveFilename (node:internal/modules/cjs/loader:1077:15)
docker-cms-1     |     at Module._load (node:internal/modules/cjs/loader:922:27)
docker-cms-1     |     at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
docker-cms-1     |     at MessagePort.<anonymous> (node:internal/main/worker_thread:164:24)
docker-cms-1     |     at [nodejs.internal.kHybridDispatch] (node:internal/event_target:737:20)
docker-cms-1     |     at exports.emitMessage (node:internal/per_context/messageport:23:28)
docker-cms-1     | Emitted 'error' event on Worker instance at:
docker-cms-1     |     at [kOnErrorMessage] (node:internal/worker:300:10)
docker-cms-1     |     at [kOnMessage] (node:internal/worker:311:37)
docker-cms-1     |     at MessagePort.<anonymous> (node:internal/worker:212:57)
docker-cms-1     |     at [nodejs.internal.kHybridDispatch] (node:internal/event_target:737:20)
docker-cms-1     |     at exports.emitMessage (node:internal/per_context/messageport:23:28) {
docker-cms-1     |   code: 'MODULE_NOT_FOUND',
docker-cms-1     |   requireStack: []
docker-cms-1     | }
docker-cms-1     |
docker-cms-1     | Node.js v18.16.1
```

My directus installation runs within docker with the following image:
```
directus/directus:10.4.3
```

And here is a repo with a fresh extension I created to isolate this from everything which is already part of my existing extension:
https://github.com/beniw/test-file-upload

It works fine as long as there is only the console.log but as soon as I add the storage import I end up with the error message above.

**Reply by beniw** [reply]
I finally had time for some further investigation and this seems to be related to the Pino logger.

It seems like Rollup can't bundle "lib/worker.js" from the Pino thread-stream dependency

https://github.com/pinojs/pino/issues/1157

I wonder if there is a way to fix this from within the extension?
